org.eclipse.xtext.web.example.entities/src/org/eclipse/xtext/web/example/entities/formatting2/EntitiesFormatter.java

```diff
@@ -28,68 +28,63 @@ protected void _format(Entities entities, @Extension IFormattableDocument docume
-		document.append(open, IHiddenRegionFormatter::newLine);
-		document.interior(open, close, IHiddenRegionFormatter::indent);
-		this.format(entity.getSuperType(), document);
+		ISemanticRegion open = regionFor(entity).keyword("{");
+		ISemanticRegion close = regionFor(entity).keyword("}");
+		document.surround(regionFor(entity).feature(DomainmodelPackage.Literals.ABSTRACT_ELEMENT__NAME),
+				this::oneSpace);
+		document.surround(entity.getSuperType(), this::oneSpace);
+		document.append(open, this::newLine);
+		document.interior(open, close, this::indent);
+		format(entity.getSuperType(), document);
 		for (Feature feature : entity.getFeatures()) {
 			document.format(feature);
 			document.append(feature, (IHiddenRegionFormatter it) -> it.setNewLines(1, 1, 2));
 		}
 	}
 
 	protected void _format(Property property, IFormattableDocument document) {
-		document.surround(textRegionExtensions.regionFor(property).keyword(":"), IHiddenRegionFormatter::noSpace);
+		document.surround(regionFor(property).keyword(":"), this::noSpace);
 		document.format(property.getType());
 	}
 
 	protected void _format(Operation operation, IFormattableDocument document) {
-		document.append(textRegionExtensions.regionFor(operation).keyword("op"), IHiddenRegionFormatter::oneSpace);
-		document.surround(textRegionExtensions.regionFor(operation).keyword("("), IHiddenRegionFormatter::noSpace);
+		document.append(regionFor(operation).keyword("op"), this::oneSpace);
+		document.surround(regionFor(operation).keyword("("), this::noSpace);
 		if (!operation.getParams().isEmpty()) {
-			for (ISemanticRegion comma : textRegionExtensions.regionFor(operation).keywords(",")) {
-				document.append(document.prepend(comma, IHiddenRegionFormatter::noSpace),
-						IHiddenRegionFormatter::oneSpace);
+			for (ISemanticRegion comma : regionFor(operation).keywords(",")) {
+				document.append(document.prepend(comma, this::noSpace), this::oneSpace);
 			}
 			for (JvmFormalParameter params : operation.getParams()) {
 				document.format(params);
 			}
-			document.prepend(textRegionExtensions.regionFor(operation).keyword(")"), IHiddenRegionFormatter::noSpace);
+			document.prepend(regionFor(operation).keyword(")"), this::noSpace);
 		}
 		if (operation.getType() != null) {
-			document.append(textRegionExtensions.regionFor(operation).keyword(")"), IHiddenRegionFormatter::noSpace);
-			document.append(document.prepend(operation.getType(), IHiddenRegionFormatter::noSpace),
-					IHiddenRegionFormatter::oneSpace);
+			document.append(regionFor(operation).keyword(")"), this::noSpace);
+			document.append(document.prepend(operation.getType(), this::noSpace), this::oneSpace);
 			document.format(operation.getType());
 		} else {
-			document.append(textRegionExtensions.regionFor(operation).keyword(")"), IHiddenRegionFormatter::oneSpace);
+			document.append(regionFor(operation).keyword(")"), this::oneSpace);
 		}
 		document.format(operation.getBody());
```

Comment:
if you expand 5 lines you will see the if else dispatching i dislike and why i want to have the option to polymorphic dispatch here

Reply:
I see - the concern is about the duality of xtend-compiled code and reflective polymorphic dispatching?

Reply:
yes. unfortunately it is not possible to call Grandparent.super.doit(xxxx)

Reply:
We could make a final protected method superDoIt() in the immediate subclass. Not pretty, though.

Reply:
that would require it to be copy&pasted.
this is why extracted the method in xtext-core